src/lib.rs

```diff
@@ -11,6 +14,60 @@ use ssb_validate::{
 };
 use ssb_verify_signatures::{par_verify_message_values, verify_message_value};
 
+// custom `enum` to allow for type conversion from js
+enum HmacKey {
+    Buf(JSArrayBuffer),
+    Str(String),
+}
+
+// implement type conversion for our custom `HmacKey` enum
+// we're primarily interested in strings and array buffers
+impl JSValue<'_> for HmacKey {
+    fn convert_to_rust(env: &JsEnv, n_value: napi_value) -> Result<Self, NjError> {
+        if let Ok(string_value) = env.convert_to_rust::<String>(n_value) {
+            Ok(Self::Str(string_value))
+        } else if let Ok(buffer_value) = env.convert_to_rust::<JSArrayBuffer>(n_value) {
+            Ok(Self::Buf(buffer_value))
+        } else {
+            Err(NjError::Other(
+                "hmacKey must be of type string, array buffer, null or undefined".to_owned(),
+            ))
+        }
+    }
+}
+
+// the `Ok()` variant for `Result` represents a valid hmac key value as a byte vector
+fn is_valid_hmac_key(hmac_key: HmacKey) -> Result<Option<Vec<u8>>, String> {
+    match hmac_key {
+        HmacKey::Buf(hmac) => {
+            let key = NetworkKey::from_slice(&hmac);
```

Comment:
Question, I don't know much about NetworkKey, but I thought that the hmac key is *not* used as the network caps (that's something that became clear to me once you commented about it in Element), but here we're still using NetworkKey, is it because both of them are 32 byte length and that's all that NetworkKey checks? I mean, maybe it could be renamed from `NetworkKey` to `KeyWith32Bytes`. I'm not suggesting that you do the rewrite, I'm just suggesting that we should either put a comment here (to clarify that the key is not a "real NetworkKey") or maybe it actually is a NetworkKey

Reply:
Yeahh, this is unfortunately a major point of confusion. The network caps key (sometimes also know as the network identifier or app key) _is_ an HMAC (hash-based message authentication code). Scuttlebutt employs this particular kind of hash-based code in two places: the secret handshake and (optional) message signing. In hindsight, they may have been better named as `network_hmac` and `message_hmac`.

I imagine the Sunrise Choir chose to call this struct `NetworkKey` because they were only thinking about the secret handshake and not the optional sign-message-with-HMAC feature. Since the cryptographic functions are the same for the secret-handshake HMAC and the message signing HMAC, we simply use the preexisting `NetworkKey` struct for convenience (it exposes the `authenticate` method which allows us to verify that the provided message bytes were indeed signed with the given HMAC).

> I mean, maybe it could be renamed from NetworkKey to KeyWith32Bytes.

One thing I could do, in addition to adding a clarifying comment, is to import the struct under an alias:

```rust
use ssb_crypto::NetworkKey as HmacMsgKey;
```

Then all references to `NetworkKey` in the code will become `HmacMsgKey`. Perhaps that will help to avoid confusion among devs who interact with this code in the future?

Reply:
Yes and yes to everything you said! Great idea with the import rename.

> In hindsight, they may have been better named as network_hmac and message_hmac

This is very insightful, I learned something new!